File: pgml-sdks/pgml/javascript/examples/summarizing_question_answering.js
const pgml = require("pgml");
require("dotenv").config();

const main = async () => {
  // Initialize the collection
  const collection = pgml.newCollection("my_javascript_sqa_collection");

  // Add a pipeline
  const model = pgml.newModel();
  const splitter = pgml.newSplitter();
  const pipeline = pgml.newPipeline(
    "my_javascript_sqa_pipeline",
    model,
    splitter,
  );
  await collection.add_pipeline(pipeline);

  // Upsert documents, these documents are automatically split into chunks and embedded by our pipeline
  const documents = [
    {
      id: "Document One",
      text: "PostgresML is the best tool for machine learning applications!",
    },
    {
      id: "Document Two",
      text: "PostgresML is open source and available to everyone!",
    },
  ];
  await collection.upsert_documents(documents);

  const query = "What is the best tool for machine learning?";

  // Perform vector search
  const queryResults = await collection
    .query()
    .vector_recall(query, pipeline)
    .limit(1)
    .fetch_all();

  // Construct context from results
  const context = queryResults
    .map((result) => {
      return result[1];
    })
    .join("\n");

  // Query for summarization
  const builtins = pgml.newBuiltins();
  const answer = await builtins.transform(
    { task: "summarization", model: "sshleifer/distilbart-cnn-12-6" },
    [context],
  );

  // Archive the collection
  await collection.archive();
  return answer;
};

main().then((results) => {
  console.log("Question summary: \n", results);
});

File: pgml-sdks/pgml/python/examples/summarizing_question_answering.py
from pgml import Collection, Model, Splitter, Pipeline, Builtins
import json
from datasets import load_dataset
from time import time
from dotenv import load_dotenv
from rich.console import Console
import asyncio


async def main():
    load_dotenv()
    console = Console()

    # Initialize collection
    collection = Collection("squad_collection")

    # Create and add pipeline
    pipeline = Pipeline(
        "squadv1",
        {
            "text": {
                "splitter": {"model": "recursive_character"},
                "semantic_search": {"model": "intfloat/e5-small"},
            }
        },
    )
    await collection.add_pipeline(pipeline)

    # Prep documents for upserting
    data = load_dataset("squad", split="train")
    data = data.to_pandas()
    data = data.drop_duplicates(subset=["context"])
    documents = [
        {"id": r["id"], "text": r["context"], "title": r["title"]}
        for r in data.to_dict(orient="records")
    ]

    # Upsert documents
    await collection.upsert_documents(documents[:200])

    # Query for answer
    query = "Who won more than 20 grammy awards?"
    console.print("Querying for context ...")
    start = time()
    results = await collection.vector_search(
        {"query": {"fields": {"text": {"query": query}}}, "limit": 3}, pipeline
    )
    end = time()
    console.print("\n Results for '%s' " % (query), style="bold")
    console.print(results)
    console.print("Query time = %0.3f" % (end - start))

    # Construct context from results
    chunks = [r["chunk"] for r in results]
    context = "\n\n".join(chunks)

    # Query for summary
    builtins = Builtins()
    console.print("Querying for summary ...")
    start = time()
    summary = await builtins.transform(
        {"task": "summarization", "model": "sshleifer/distilbart-cnn-12-6"},
        [context],
    )
    end = time()
    console.print("Summary '%s'" % summary, style="bold")
    console.print("Query time = %0.3f" % (end - start))

    # Archive collection
    await collection.archive()


if __name__ == "__main__":
    asyncio.run(main())
